### DISCUSSION TITLE
Missing safepath.Join for darwin

### DISCUSSION THREAD

**Discussion by matejvasek** [moby/moby#47750] [General]
It appears that `safepath.Join()` is missing for darwin. There is however a Linux version. Would this Linux version work also for darwin.

**Comment by matejvasek** [comment]
Could we just rename `internal/safepath/join_linux.go` to `internal/safepath/join_unix.go` and put appropriate build tags to the file?

**Comment by matejvasek** [comment]
cc @vvoland

**Comment by matejvasek** [comment]
We use docker as a library in some project, I think we do not actually invoke `safepath.Join()` however absence of definition breaks compilation for us. So even implementation that just `panic()` is OK for us.

**Reply by thaJeztah** [reply]
Do you have more context on what parts you're using? It looks like it's used in the volumes/mounts` package, which is used by the daemon code (which on itself is only implemented for Linux and Windows).

For context; we're working on moving the codebase to become a Go module, and want to be more explicit on "internal" and "library" code; most parts of the `daemon/` code (and packages consumed by it) are considered "internal" (although not currently hard restricted).

**Reply by matejvasek** [reply]
The https://github.com/buildpacks/pack project should be using only client.

**Reply by matejvasek** [reply]
As I said I think `safepath.Join()` is not really invoked by the pack project. However it prevents compilation.

**Reply by matejvasek** [reply]
The only place where mounts pkg is probably used here: https://github.com/buildpacks/pack/blob/72414e5952a89d7b43270815cd96aec9737eeb7c/pkg/client/build.go#L1382-L1389

**Reply by matejvasek** [reply]
Removing that fixes the compilation. But it would be nice to be able to use existing parsing logic from docker project. Otherwise we need our own parsing logic.

**Reply by thaJeztah** [reply]
Thanks for the extra context, that's definitely useful! There's still many decisions to be made on how to slice/dice packages into "public" and "internal". We've encountered various cases where packages ended up in unexpected places; sometimes that was just "incorrect use", but in some cases it was due to some implicit dependency between them (so some refactoring needed). We're also aware that there's still a lot of logic implemented in the CLI itself; not all of which being easy to implement (and code we should move to be more easy to consume).

We'd probably need to have a closer look at the code you linked to get an understanding of why/how it's used exactly, and if we need to refactor things to either _allow_ that code to be more officially or to split relevant parts separate.

For the immediate issue, I'm not against implementing a stub that `panics`; probably accompanied by a code-comment outlining "why" (and that it may not be _permanent_ - so there could be a TODO and a tracking ticket for it). I'm mostly trying to avoid implementing a "partial" (seems to be functional) implementation to not give any assumption that it's actually implemented. We've had such cases in the past where implementations were contributed (which turned out to be unused, and only to make it compile), and we ended up trying to keep them functional (but there not being any actual consumers).

Maybe @vvoland also has some ideas; I haven't looked at the code yet to see if it's possible to split out relevant parts to a separate package (allowing parsing to still work, without the securejoin parts)


I'm also curious if there's code in the CLI that would be suitable for this purpose; I recall there's parsing code there as well, but maybe it's not yet easy to consume; from a quick search;
https://github.com/docker/cli/blob/865190615ba80be5a9ebc21ffcde281ddc36bb5b/cli/command/container/opts.go#L371-L375
https://github.com/docker/cli/blob/865190615ba80be5a9ebc21ffcde281ddc36bb5b/cli/compose/loader/volume.go#L16

**Reply by matejvasek** [reply]
@jjbustamante maybe `loader.ParseVolume()` could work for us?

**Reply by matejvasek** [reply]
@thaJeztah thanks for reply, maybe `loader.ParseVolume()` is the way for us.

**Reply by thaJeztah** [reply]
I only had a quick peek at the code in the CLI, so be sure to have a look around in the code to see if that's the correct implementation to use.

(in all honesty, handling of volumes is in dire need of a bit of a cleanup, as there's multiple (legacy) APIs still in use for backward-compatibility); let me know if you managed to find a solution though!

**Reply by vvoland** [reply]
It won't work, the issue is that the parser returns `MountPoint`, which is defined in `volumes/mounts/mount.go`, which also uses the safepath 🙈 

I don't see an easy way to keep the build failure when actually building the daemon for non-supported platforms with how these things are structured right now... So the easiest way to unblock this is adding a `_other.go` safepath implementation that just returns an error (no need for panic I think).

**Reply by jjbustamante** [reply]
Hi @thaJeztah , @vvoland for now I worked around our problem on `darwin` implementing our validation using `loader.ParseVolume()` and some string hacks, really awful but it unblock us. 
If `_other.go` safepath implementation can be added that will be great! I will keep and eye on this.
